etl2574: handling camus offset out of range AND etl2573: handling error in KafkaReader.fetch()

zliu41 opened the pull request:
If out of range, skip topic and report job failure.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/KafkaReader.java:
```diff
@@ -163,53 +168,91 @@ public boolean fetch() throws IOException {
+    } catch (Exception e) {
+      log.info("Exception generated during fetch for topic " + kafkaRequest.getTopic()
+          + ". This topic will be skipped.");
+      e.printStackTrace();
```
Don't use e.printStackTrace()

zliu41 replied:
Fixed. The printStackTrace() is in the old code, and there are other occurrences of printStackTrace() in the old code.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/KafkaReader.java:
```diff
@@ -163,53 +168,91 @@ public boolean fetch() throws IOException {
-          messageIter = null;
-          return false;
+  private void refreshTopicMetadata() {
+    TopicMetadataRequest request = new TopicMetadataRequest(Collections.singletonList(kafkaRequest.getTopic()));
```
Should be topic + partition.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/KafkaReader.java:
```diff
@@ -163,53 +168,91 @@ public boolean fetch() throws IOException {
+    }
+  }
+
+  private boolean processFetchResponse(FetchResponse fetchResponse, long tempTime) {
```
Is this the same as the old code?  How come it looks like it has more lines.

zliu41 replied:
Line 215-249 is the same as old code line 171-205. I added a try-catch outside it, because there's a try catch in old code 207-211.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/CamusJob.java:
```diff
@@ -390,6 +390,10 @@ public int compare(FileStatus f1, FileStatus f2) {
       throw new RuntimeException("Camus saw errors, check stderr");
     }
+
+    if (EtlInputFormat.reportJobFailureDueToOffsetOutOfRange) {
```
It's also better to introduce a config parameter to decide whether to throw Exception on this case.

Also How did you test all these error cases?

HenryCaiHaiying replied:
The new code looks good.

zliu41 commented:
1. Added unit test for etl2574: offset out of range
 - added method "testJobFailureDueToOffsetOutOfRange()" in CamusJobTest.java
 - added class "EtlInputFormatFake", the difference from "EtlInputFormat" is that the fake class always return out-of-range offsets.
 - added field "useFakeEtlInputFormatForUnitTest" in CamusJob class. If ==true, it will load EtlInputFormatFake instead of EtlInputFormat.

2. Added unit test for etl2573: fetch() failure
 - Use easymock for kafkaReader.simpleConsumer

3. Added config parameter "etl.fail.on.offset.outofrange".

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/KafkaReader.java:
```diff
@@ -163,53 +168,96 @@ public boolean fetch() throws IOException {
     try {
       fetchResponse = simpleConsumer.fetch(fetchRequest);
       if (fetchResponse.hasError()) {
+        throw new RuntimeException();
```
The exception should have some error messages.

zliu41 replied:
This exception is solely for consumption in line 174 (so that we don't need to write duplicate code for fetchResponse=null, fetchResponse.hasError(), and fetch() has exception. All three cases will go to line 174). The error message is printed in line 175.

HenryCaiHaiying replied:
I think it's better to put the fetchResponse.error in the exception message so the caller will have more information to print.

zliu41 replied:
Fixed.

HenryCaiHaiying commented on camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormatFake.java:
```diff
@@ -0,0 +1,595 @@
+package com.linkedin.camus.etl.kafka.mapred;
```
Can you mock the ETLInputFormat?  There are too many code duplications here.

zliu41 replied:
Fixed.

HenryCaiHaiying commented on camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/common/KafkaReaderTest.java:
```diff
@@ -0,0 +1,51 @@
+    IllegalArgumentException, IllegalAccessException, IOException {
+    SimpleConsumer consumer = EasyMock.createNiceMock(SimpleConsumer.class);
+    EasyMock.expect(consumer.fetch((FetchRequest) EasyMock.anyObject())).andReturn(null);
+    EasyMock.expectLastCall().times(2);
```
Why the fetch will be called two times?

zliu41 replied:
One for the first attempt, and one for the retry, after refreshing topic metadata.

HenryCaiHaiying replied:
OK.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/KafkaReader.java:
```diff
@@ -163,53 +168,96 @@ public boolean fetch() throws IOException {
+      return processFetchResponse(fetchResponse, tempTime);
+    } catch (Exception e) {
+      log.info("Exception generated during fetch for topic " + kafkaRequest.getTopic()
+          + ". Will refresh topic metadata and retry.");
```
The exception 'e' is not printed in the log.

zliu41 replied:
Fixed.

zliu41 commented:
Modified unit test for offset-out-of-range. Instead of using a fake EtlInputFormatFake class which has duplicate code, now using a flag (useMockRequestForUnitTest) in EtlInputFormat.java. If flag==true, EtlInputFormat will replace the actual CamusRequest with a mocked request, which contains out-of-range offsets.

Also fixed other minor issues.

HenryCaiHaiying commented on camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/CamusJobTest.java:
```diff
@@ -140,6 +148,32 @@ public void runJob() throws Exception {
+    createMockRequestOffsetTooEarly();
+    try {
+      job.run();
+      fail("Should have thrown RuntimeException due to offset out of range.");
```
Do you need to call EasyMock.verify somewhere?

HenryCaiHaiying commented:
Looks good to be merged.

zliu41 commented:
Rewrote unit test to use the new mock framework.
